pkg/scheduler/framework/framework.go

```diff
@@ -143,8 +174,738 @@ func (f *framework) EventRecorder() record.EventRecorder {
+	scoredClusters, err := f.runScorePlugins(ctx, state, policy, passed)
+	if err != nil {
+		klog.ErrorS(err, "failed to run score plugins", "schedulingPolicySnapshot", schedulingPolicySnapshotRef)
+		return ctrl.Result{}, controller.NewUnexpectedBehaviorError(err)
+	}
+
+	// Pick the top scored clusters.
+	klog.V(2).InfoS("Picking top-scored clusters", "policy", schedulingPolicySnapshotRef)
+
+	// Calculate the number of clusters to pick.
+	numOfClustersToPick := calcNumOfClustersToSelect(state.desiredBatchSize, state.batchSizeLimit, len(scoredClusters))
+
+	// Do a sanity check; normally this branch will never run, as earlier check
+	// guarantees that the number of clusters to pick is always no greater than number of
+	// scored clusters.
+	if numOfClustersToPick > len(scoredClusters) {
+		err := fmt.Errorf("number of clusters to pick is greater than number of scored clusters: %d > %d", numOfClustersToPick, len(scoredClusters))
+		klog.ErrorS(err, "failed to calculate number of clusters to pick", "schedulingPolicySnapshot", schedulingPolicySnapshotRef)
+		return ctrl.Result{}, controller.NewUnexpectedBehaviorError(err)
+	}
+
+	// Pick the clusters.
+	//
+	// Note that at this point of the scheduling cycle, any cluster associated with a currently
+	// active or creating binding should be filtered out already.
+	picked := pickTopNScoredClusters(scoredClusters, numOfClustersToPick)
+
+	// Cross-reference the newly picked clusters with obsolete bindings; find out
+	//
+	// * bindings that should be created, i.e., create a binding for every cluster that is newly picked
+	//   and does not have a binding associated with;
+	// * bindings that should be updated, i.e., associate a binding whose target cluster is picked again
+	//   in the current run with the latest score;
+	// * bindings that should be deleted, i.e., mark a binding as deleting if its target cluster is no
+	//   longer picked in the current run.
+	//
+	// Fields in the returned bindings are fulfilled and/or refreshed as applicable.
+	klog.V(2).InfoS("Cross-referencing bindings with picked clusters", "schedulingPolicySnapshot", schedulingPolicySnapshotRef)
+	toCreate, toUpdate, toDelete, err := crossReferencePickedCustersAndObsoleteBindings(crpName, policy, picked, obsolete)
+
+	// Manipulate bindings accordingly.
+	klog.V(2).InfoS("Creating, updating, and/or deleting bindings", "policy", schedulingPolicySnapshotRef)
+
+	// Create new bindings.
+	if err := f.createBindings(ctx, toCreate); err != nil {
+		klog.ErrorS(err, "Failed to create new bindings", "schedulingPolicySnapshot", schedulingPolicySnapshotRef)
+		return ctrl.Result{}, err
+	}
+
+	// Update existing bindings.
+	//
+	// Note that a race condition may arise here, when a rollout controller attempts to update bindings
+	// at the same time with the scheduler. An error induced requeue will happen in this case.
+	if err := f.updateBindings(ctx, toUpdate); err != nil {
```

Comment:
is there real race between scheduler/rollout controller in this case as they manage different part of the binding?

Reply:
Hi Ryan! A race could happen if the rollout controller are transitioning a binding from creating stated to active while the scheduler is trying to update the binding at the same time.